File: Competitive Programming 3/Section 1.4/Harder (mode tedious)/227 - Puzzle/C++/main.cpp
#include<bits/stdc++.h>

using namespace std;

char grid[5][5];
int iE,jE;
string line;
map<char,vector<int> >getAdjacent={
    make_pair('L',vector<int>{0,-1}),
    make_pair('R',vector<int>{0,1}),
    make_pair('B',vector<int>{1,0}),
    make_pair('A',vector<int>{-1,0})
};
bool valid(int i,int j){
    return i>-1&&j>-1&&i<5&&j<5;
}
bool successfull(char command){
    auto it = getAdjacent.find(command);
    if(it==getAdjacent.end())
        return false;
    auto adjMov=it->second;
    auto newIE=iE+adjMov[0],newJE=jE+adjMov[1];
    if(!valid(newIE,newJE))
        return false;
    //make move
    grid[iE][jE]=grid[newIE][newJE];
    grid[newIE][newJE]=' ';
    iE=newIE,jE=newJE;
    return true;
}
int main(){
    ios_base::sync_with_stdio(0);cin.tie(0);
    for(int cs=1;1;cs++){
        for(int i=0;i<5;i++){
            getline(cin,line);
            if(line[0]=='Z')
                return 0;
            for(int j=0;j<5;j++){
                grid[i][j]=line[j];
                if(grid[i][j]==' ')
                    iE=i,jE=j;
            }
        }
        cout<<(cs>1?"\n":"")<<"Puzzle #"<<cs<<":\n";
        auto possible=true,moreCommands=true;
        while(moreCommands){
            getline(cin,line);
            for(auto command:line){
                if(command=='0'){
                    moreCommands=false;
                    break;
                }else if(possible&&!successfull(command)){
                    possible=false;
                    cout<<"This puzzle has no final configuration."<<"\n";
                }
            }
        }
        if(possible)
            for(int i=0;i<5;i++){
                for(int j=0;j<5;j++)
                    cout<<(j>0?" ":"")<<grid[i][j];
                cout<<"\n";
            }
    }
    return 0;
}

File: Competitive Programming 3/Section 1.4/Harder (mode tedious)/227 - Puzzle/Python/main.py
grid,iE,jE=[['\0' for _ in range(5)] for _ in range(5)],int(),int()
getAdjacent={
    'L':[0,-1],
    'R':[0,1],
    'B':[1,0],
    'A':[-1,0]
}
def valid(i,j):
    return i>-1 and j>-1 and i<5 and j<5
def successfull(command):
    global iE,jE
    adjMov=getAdjacent.get(command)
    if not adjMov:
        return False
    newIE,newJE=iE+adjMov[0],jE+adjMov[1]
    if not valid(newIE,newJE):
        return False
    #make move
    grid[iE][jE]=grid[newIE][newJE]
    grid[newIE][newJE]=' '
    iE,jE=newIE,newJE
    return True
def main():
    global iE,jE
    cs=1
    while(True):
        for i in range(5):
            line=input()
            if line[0]=='Z':
                return
            for j in range(5):
                grid[i][j]=line[j]
                if grid[i][j]==' ':
                    iE,jE=i,j
        print(('\n' if cs>1 else '')+'Puzzle #'+str(cs)+':')
        possible,moreCommands=True,True
        while(moreCommands):
            for command in input():
                if command=='0':
                    moreCommands=False
                    break
                elif possible and not successfull(command):
                    possible=False
                    print("This puzzle has no final configuration.")
        if possible:
            for row in grid:
                for i,el in enumerate(row):
                    print((' ' if i>0 else '')+el,end='')
                print()
        cs+=1
main()